lib/google/gax/grpc/status_details.rb

```diff
@@ -0,0 +1,60 @@
+# Copyright 2016, Google Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+#     * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#     * Neither the name of Google Inc. nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+require "grpc"
+require "grpc/google_rpc_status_utils"
+require "google/gax/errors"
+require "google/protobuf/well_known_types"
+# Required in order to deserialize common error detail proto types
+require "google/rpc/error_details_pb"
+
+module Google
+  module Gax
+    class GaxError < StandardError
+      def status_details
+        return unless cause.is_a? GRPC::BadStatus
+
+        rpc_status = GRPC::GoogleRpcStatusUtils.extract_google_rpc_status cause.to_status
+
+        return if rpc_status.nil?
+
+        rpc_status.details.map do |detail|
+          begin
+            detail_type = Google::Protobuf::DescriptorPool.generated_pool.lookup detail.type_name
+            detail.unpack detail_type.msgclass
+          rescue StandardError
+            detail
+          end
+        end
+      rescue StandardError
+        nil
```

Comment:
Is this meant to replace the previous error message about malformed responses? Can we do any better than returning `nil` here?

Reply:
We can raise an error. I didn't understand why the previous implementation would return a string when malformed, instead of allowing the `Google::Protobuf::ParseError` to be raised. Having to check for a single string instead of an array is not significantly different than handling an error.

Reply:
Yeah, the old behavior was confusing (we had a recently complain from Sam actually) so that sounds more reasonable to me.

Reply:
I would prefer to raise an error, or in general not to drop the error on the floor.

There are other cases where `nil` is returned. Since the happy path for this method returns an array, can we have all code paths either return an array (even an empty one) or raise an error? It's weird to return a collection or nil; it means callers can't just call enumerable methods on the result like they might want to, but have to add nil checks.

Reply:
We discussed raising or not raising in grpc/grpc#18752. My view here is that this is a convenience method, and it is more convenient to not have to add error handling when trying to pull values from the object.

FWIW, I do agree that it would be nicer to return an empty array here instead of nil. Should I make that change?

Reply:
The justification for returning `nil` here is that it matches the behavior of the other error methods.

```ruby
      def code
        return nil unless cause&.respond_to? :code
        cause.code
      end

      def details
        return nil unless cause&.respond_to? :details
        cause.details
      end

      def metadata
        return nil unless cause&.respond_to? :metadata
        cause.metadata
      end
```

Reply:
Yeah I think I'd prefer to see empty arrays/hashes for all those. (i.e. if the cause conceptually has no status details, then return an empty array meaning no details.) For `code` I'm not sure. If that's numeric, then maybe `nil` is reasonable to indicate absence of value.

Reply:
The rest of this PR LGTM. I think I'm just stuck on the return values for the convenience methods. Especially if they're intended as convenience methods they should be... convenient.

Reply:
@dazuma Updated. PTAL.